Simplenote/src/main/java/com/automattic/simplenote/DeepLinkActivity.kt

```diff
@@ -0,0 +1,73 @@
+package com.automattic.simplenote
+
+import android.content.Intent
+import android.net.Uri
+import android.os.Bundle
+import androidx.appcompat.app.AppCompatActivity
+import com.automattic.simplenote.authentication.SimplenoteAuthenticationActivity
+import com.automattic.simplenote.utils.AuthUtils
+import com.automattic.simplenote.utils.IntentUtils
+import net.openid.appauth.RedirectUriReceiverActivity
+
+class DeepLinkActivity : AppCompatActivity() {
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        val uri = intent.data
+        when (uri?.host) {
+            AUTHENTICATION_SCHEME -> {
+                val intent = Intent(this, RedirectUriReceiverActivity::class.java)
+                intent.setData(uri)
+                startActivity(intent)
+            }
+            VERIFIED_WEB_SCHEME -> {
+                // New MagicLink
+                startMagicLinkConfirmation(uri)
+            }
+            LOGIN_SCHEME -> {
+                if (queryParamContainsData(uri.query, AUTH_KEY_QUERY) && queryParamContainsData(uri.query, AUTH_CODE_QUERY)) {
+                    startMagicLinkConfirmation(uri)
+                } else {
+                    val intent = IntentUtils.maybeAliasedIntent(applicationContext)
+                    val app = application as Simplenote
+                    val email = AuthUtils.extractEmailFromMagicLink(uri)
+                    if (app.isLoggedIn &&
+                        email.lowercase() != app.userEmail.lowercase()
+                    ) {
+                        intent.putExtra(NotesActivity.KEY_ALREADY_LOGGED_IN, true)
+                    } else {
+                        AuthUtils.magicLinkLogin(application as Simplenote, uri)
+                    }
+                    intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK or Intent.FLAG_ACTIVITY_NEW_TASK)
+                    startActivity(intent)
+                }
+            }
+        }
+        finish()
+    }
+
+    private fun startMagicLinkConfirmation(uri: Uri?) {
+        val authKey = uri?.getQueryParameter(AUTH_KEY_QUERY)
+        val authCode = uri?.getQueryParameter(AUTH_CODE_QUERY)
+        if (!authKey.isNullOrBlank() && !authCode.isNullOrBlank()) {
+            val intent = Intent(this, SimplenoteAuthenticationActivity::class.java)
+            intent.putExtra(SimplenoteAuthenticationActivity.KEY_IS_MAGIC_LINK, true);
+            intent.putExtra(SimplenoteAuthenticationActivity.KEY_MAGIC_LINK_AUTH_KEY, authKey)
+            intent.putExtra(SimplenoteAuthenticationActivity.KEY_MAGIC_LINK_AUTH_CODE, authCode)
+            intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK or Intent.FLAG_ACTIVITY_NEW_TASK)
+            startActivity(intent)
+        }
```

Comment:
I think we can add a Toast message here for the else case, saying that the link was invalid. Just use the app context to launch the Toast.

Reply:
And take the user to the notes activity.

Reply:
Assuming no other big issues, I can make an update to this on a subsequent PR if that's ok.

Reply:
Yeah, this is a pretty edge case, but we cannot be sure if the email client will mess up the link or something like that. So if the link is wrong, then we need to send the user to the login screen if it is not logged in.

Reply:
> So if the link is wrong, then we need to send the user to the login screen if it is not logged in.

Oh. I see what you mean 🤔 . Fair enough. Created the issue.